### PR TITLE
build: Fix qt build for riscv64

### LINKED ISSUE #29: build: Building depends for the `riscv64-linux-gnu` host fails
On the main branch (2aa44e608ac70ed7e56994634d8c15cf192f8390):
```
$ env HOSTS='riscv64-linux-gnu' ./contrib/guix/guix-build
...
linking ../../bin/qml
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L567':
qv4typedarray.cpp:(.text._Z9atomicXorIhEyPcN3QV45ValueE[_Z9atomicXorIhEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_xor_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L590':
qv4typedarray.cpp:(.text._Z9atomicXorItEyPcN3QV45ValueE[_Z9atomicXorItEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_xor_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L613':
qv4typedarray.cpp:(.text._Z9atomicSubIhEyPcN3QV45ValueE[_Z9atomicSubIhEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_sub_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L636':
qv4typedarray.cpp:(.text._Z8atomicOrItEyPcN3QV45ValueE[_Z8atomicOrItEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_or_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L659':
qv4typedarray.cpp:(.text._Z9atomicAddIaEyPcN3QV45ValueE[_Z9atomicAddIaEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_add_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L682':
qv4typedarray.cpp:(.text._Z8atomicOrIaEyPcN3QV45ValueE[_Z8atomicOrIaEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_or_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L705':
qv4typedarray.cpp:(.text._Z14atomicExchangeIaEyPcN3QV45ValueE[_Z14atomicExchangeIaEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_exchange_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L728':
qv4typedarray.cpp:(.text._Z8atomicOrIsEyPcN3QV45ValueE[_Z8atomicOrIsEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_or_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L751':
qv4typedarray.cpp:(.text._Z9atomicXorIaEyPcN3QV45ValueE[_Z9atomicXorIaEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_xor_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L774':
qv4typedarray.cpp:(.text._Z9atomicAndItEyPcN3QV45ValueE[_Z9atomicAndItEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_and_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L797':
qv4typedarray.cpp:(.text._Z9atomicSubIaEyPcN3QV45ValueE[_Z9atomicSubIaEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_sub_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L842':
qv4typedarray.cpp:(.text._Z9atomicXorIsEyPcN3QV45ValueE[_Z9atomicXorIsEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_xor_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L929':
qv4typedarray.cpp:(.text._Z9atomicAddIhEyPcN3QV45ValueE[_Z9atomicAddIhEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_add_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L974':
qv4typedarray.cpp:(.text._Z14atomicExchangeIhEyPcN3QV45ValueE[_Z14atomicExchangeIhEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_exchange_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L997':
qv4typedarray.cpp:(.text._Z9atomicAndIaEyPcN3QV45ValueE[_Z9atomicAndIaEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_and_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1020':
qv4typedarray.cpp:(.text._Z8atomicOrIhEyPcN3QV45ValueE[_Z8atomicOrIhEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_or_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1064':
qv4typedarray.cpp:(.text._Z9atomicAndIhEyPcN3QV45ValueE[_Z9atomicAndIhEyPcN3QV45ValueE]+0x72): undefined reference to `__atomic_fetch_and_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1109':
qv4typedarray.cpp:(.text._Z14atomicExchangeItEyPcN3QV45ValueE[_Z14atomicExchangeItEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_exchange_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1132':
qv4typedarray.cpp:(.text._Z9atomicSubItEyPcN3QV45ValueE[_Z9atomicSubItEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_sub_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1155':
qv4typedarray.cpp:(.text._Z9atomicAddIsEyPcN3QV45ValueE[_Z9atomicAddIsEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_add_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1200':
qv4typedarray.cpp:(.text._Z14atomicExchangeIsEyPcN3QV45ValueE[_Z14atomicExchangeIsEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_exchange_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1223':
qv4typedarray.cpp:(.text._Z9atomicAndIsEyPcN3QV45ValueE[_Z9atomicAndIsEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_and_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1246':
qv4typedarray.cpp:(.text._Z9atomicSubIsEyPcN3QV45ValueE[_Z9atomicSubIsEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_sub_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1269':
qv4typedarray.cpp:(.text._Z9atomicAddItEyPcN3QV45ValueE[_Z9atomicAddItEyPcN3QV45ValueE]+0x74): undefined reference to `__atomic_fetch_add_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1539':
qv4typedarray.cpp:(.text._Z21atomicCompareExchangeItEyPcN3QV45ValueES2_[_Z21atomicCompareExchangeItEyPcN3QV45ValueES2_]+0xde): undefined reference to `__atomic_compare_exchange_2'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1584':
qv4typedarray.cpp:(.text._Z21atomicCompareExchangeIaEyPcN3QV45ValueES2_[_Z21atomicCompareExchangeIaEyPcN3QV45ValueES2_]+0xe2): undefined reference to `__atomic_compare_exchange_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1629':
qv4typedarray.cpp:(.text._Z21atomicCompareExchangeIhEyPcN3QV45ValueES2_[_Z21atomicCompareExchangeIhEyPcN3QV45ValueES2_]+0xde): undefined reference to `__atomic_compare_exchange_1'
riscv64-linux-gnu-ld: /bitcoin/depends/work/build/riscv64-linux-gnu/qt/5.12.11-a191f20871a/qtdeclarative/lib/libQt5Qml.a(qv4typedarray.o): in function `.L1674':
qv4typedarray.cpp:(.text._Z21atomicCompareExchangeIsEyPcN3QV45ValueES2_[_Z21atomicCompareExchangeIsEyPcN3QV45ValueES2_]+0xe2): undefined reference to `__atomic_compare_exchange_2'
collect2: error: ld returned 1 exit status
...
```

### PR DESCRIPTION
**Pull request by hebasto**
Fixes #29.

#### Guix build:
```
$ env HOSTS='riscv64-linux-gnu' ./contrib/guix/guix-build
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
9399101a4268eb78b4c13d0895964862211d87e9a12bd154c11ea4ca257f02d6  guix-build-3a0f948c0bc9/output/dist-archive/bitcoin-3a0f948c0bc9.tar.gz
c4c57780ee1d99721dc370c35cc68dfccdd5455a29c0c526a1171f3b8dfa4dc8  guix-build-3a0f948c0bc9/output/riscv64-linux-gnu/SHA256SUMS.part
4840f8a57c43e3956073681aeb3779f326fa89102a643c5bf5acd26790107284  guix-build-3a0f948c0bc9/output/riscv64-linux-gnu/bitcoin-3a0f948c0bc9-riscv64-linux-gnu-debug.tar.gz
95910c5070e14532c2b96a10662c1ae77c45199649234248ee71017e15cc7a71  guix-build-3a0f948c0bc9/output/riscv64-linux-gnu/bitcoin-3a0f948c0bc9-riscv64-linux-gnu.tar.gz
```

### CODE REVIEW

File: depends/patches/qt/fix_riscv_atomic.patch [hunk 1]
```diff
@@ -0,0 +1,17 @@
+A workaroud the error when build for riscv64.
```
Comment (jarolrod) [inline depends/patches/qt/fix_riscv_atomic.patch:1]:
```suggestion
A workaround for error when building for riscv64.
```

**Comment by jarolrod** [comment]
GUIX hashes mine match @hebasto:
``` diff
find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

00bd2214012476a91e32cddc720ef3ba12dcd992c0fea4f656f8e1ca672b1768  guix-build-bd86bb536645/output/dist-archive/bitcoin-bd86bb536645.tar.gz
1f2f80002f6bda09073f3da6b292d232e1c657ecf1e6f0c59ac01b1f5d3dd474  guix-build-bd86bb536645/output/x86_64-w64-mingw32/SHA256SUMS.part
b988b0ce195f536799b30194f573a0ae0d26a32b00c1a5b54039d62e60f92db9  guix-build-bd86bb536645/output/x86_64-w64-mingw32/bitcoin-bd86bb536645-win-unsigned.tar.gz
e89793b517eb171ecf0668f5f004af2df43077614f2cb22e0eb9af736b51dc80  guix-build-bd86bb536645/output/x86_64-w64-mingw32/bitcoin-bd86bb536645-win64-debug.zip
ccc6fc78cf0de61772bab59a83b2370aa825bd5c6d0af126b19c02f985b04fda  guix-build-bd86bb536645/output/x86_64-w64-mingw32/bitcoin-bd86bb536645-win64-setup-unsigned.exe
d753a67e614ad33d6c07e8e32183feec81dee0edd51f319e9bea698c985cf0b6  guix-build-bd86bb536645/output/x86_64-w64-mingw32/bitcoin-bd86bb536645-win64.zip
```

**Comment by hebasto** [comment]
Updated 5ed24d52e3a9c085c9cbeb9b4b43bf3bf5da7422 -> 3a0f948c0bc93b571168165c54966bf164dc356a ([pr30.02](https://github.com/hebasto/gui-qml/commits/pr30.02) -> [pr30.03](https://github.com/hebasto/gui-qml/commits/pr30.03)):
- rebased on top of the #28 
- addressed @jarolrod's [comment](https://github.com/bitcoin-core/gui-qml/pull/30#discussion_r698061571)

File: depends/patches/qt/fix_riscv_atomic.patch (hunk 1 above)
Comment (hebasto) [reply depends/patches/qt/fix_riscv_atomic.patch]: Thanks! [Updated](https://github.com/bitcoin-core/gui-qml/pull/30#issuecomment-908117764).

**Comment by jarolrod** [comment]
GUIX hashes mine match @hebasto:

```
find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

9399101a4268eb78b4c13d0895964862211d87e9a12bd154c11ea4ca257f02d6  guix-build-3a0f948c0bc9/output/dist-archive/bitcoin-3a0f948c0bc9.tar.gz
c4c57780ee1d99721dc370c35cc68dfccdd5455a29c0c526a1171f3b8dfa4dc8  guix-build-3a0f948c0bc9/output/riscv64-linux-gnu/SHA256SUMS.part
4840f8a57c43e3956073681aeb3779f326fa89102a643c5bf5acd26790107284  guix-build-3a0f948c0bc9/output/riscv64-linux-gnu/bitcoin-3a0f948c0bc9-riscv64-linux-gnu-debug.tar.gz
95910c5070e14532c2b96a10662c1ae77c45199649234248ee71017e15cc7a71  guix-build-3a0f948c0bc9/output/riscv64-linux-gnu/bitcoin-3a0f948c0bc9-riscv64-linux-gnu.tar.gz
```

File: depends/patches/qt/fix_riscv_atomic.patch [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+Patch concept: https://bugs.gentoo.org/790689
+
+
+--- old/qtdeclarative/src/qml/qml.pro
++++ new/qtdeclarative/src/qml/qml.pro
+@@ -19,6 +19,7 @@ solaris-cc*:QMAKE_CXXFLAGS_RELEASE -= -O2
+ 
+ # Ensure this gcc optimization is switched off for mips platforms to avoid trouble with JIT.
+ gcc:isEqual(QT_ARCH, "mips"): QMAKE_CXXFLAGS += -fno-reorder-blocks
++gcc:isEqual(QT_ARCH, "riscv64"): LIBS += -latomic
```
Comment (jarolrod) [inline depends/patches/qt/fix_riscv_atomic.patch:14]: This differs form the mentioned patch by excluding the check for riscv32. This is ok as we do not release riscv32 binaries:

original patch's relevant line
```diff
+ gcc:if(isEqual(QT_ARCH, "riscv32")|isEqual(QT_ARCH, "riscv64")): LIBS += -latomic
```

**Comment by jarolrod** [comment]
ACK 3a0f948c0bc93b571168165c54966bf164dc356a

Confirmed the issue on master, the supplied patches in this PR allow me to successfully build. I did not actually test the resulting binary on riscv.

**Pull request merged**